File: scripts/sync-version.js
#!/usr/bin/env node
/**
 * 版本号同步脚本
 * 从根目录 package.json 读取 version 字段，自动同步到所有相关文件。
 *
 * 用法:
 *   node scripts/sync-version.js          # 同步当前版本
 *   node scripts/sync-version.js 1.0.4    # 设置并同步到指定版本
 */
'use strict';

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');

// ─── 读取/设置版本号 ──────────────────────────
const rootPkgPath = path.join(ROOT, 'package.json');
const rootPkg = JSON.parse(fs.readFileSync(rootPkgPath, 'utf-8'));

const newVersion = process.argv[2] || rootPkg.version;
if (!newVersion || !/^\d+\.\d+\.\d+/.test(newVersion)) {
  console.error('无效的版本号:', newVersion);
  process.exit(1);
}

// 如果指定了新版本号，先更新根 package.json
if (process.argv[2] && rootPkg.version !== newVersion) {
  rootPkg.version = newVersion;
  fs.writeFileSync(rootPkgPath, JSON.stringify(rootPkg, null, 2) + '\n', 'utf-8');
  console.log(`  [ROOT] package.json → ${newVersion}`);
}

console.log(`\n同步版本号: ${newVersion}\n`);

let updatedCount = 0;

// ─── 辅助函数 ──────────────────────────────────
function replaceInFile(relPath, pattern, replacement) {
  const absPath = path.join(ROOT, relPath);
  if (!fs.existsSync(absPath)) {
    console.log(`  [SKIP] ${relPath} (文件不存在)`);
    return;
  }
  let content = fs.readFileSync(absPath, 'utf-8');
  const newContent = content.replace(pattern, replacement);
  if (content !== newContent) {
    fs.writeFileSync(absPath, newContent, 'utf-8');
    console.log(`  [OK]   ${relPath}`);
    updatedCount++;
  } else {
    console.log(`  [--]   ${relPath} (已是最新)`);
  }
}

function updateJsonFile(relPath, jsonPath) {
  const absPath = path.join(ROOT, relPath);
  if (!fs.existsSync(absPath)) {
    console.log(`  [SKIP] ${relPath} (文件不存在)`);
    return;
  }
  const obj = JSON.parse(fs.readFileSync(absPath, 'utf-8'));
  const keys = jsonPath.split('.');
  let target = obj;
  for (let i = 0; i < keys.length - 1; i++) {
    target = target[keys[i]];
  }
  const lastKey = keys[keys.length - 1];
  if (target[lastKey] !== newVersion) {
    target[lastKey] = newVersion;
    fs.writeFileSync(absPath, JSON.stringify(obj, null, 2) + '\n', 'utf-8');
    console.log(`  [OK]   ${relPath} → ${jsonPath}`);
    updatedCount++;
  } else {
    console.log(`  [--]   ${relPath} (已是最新)`);
  }
}

// ─── 执行同步 ──────────────────────────────────

// 1. frontend/package.json
updateJsonFile('frontend/package.json', 'version');

// 2. resources/config/default.json
updateJsonFile('resources/config/default.json', 'app.version');

// 3. frontend/src/config/constants.ts
replaceInFile(
  'frontend/src/config/constants.ts',
  /export const SYSTEM_VERSION = "[^"]+"/,
  `export const SYSTEM_VERSION = "${newVersion}"`
);

// 4. frontend/src/config/appConfig.ts
replaceInFile(
  'frontend/src/config/appConfig.ts',
  /version: "[^"]+"/,
  `version: "${newVersion}"`
);

// 5. frontend/src/stores/app.ts
replaceInFile(
  'frontend/src/stores/app.ts',
  /version: '[^']+'/,
  `version: '${newVersion}'`
);

// 6. frontend/src/utils/local-storage.ts
replaceInFile(
  'frontend/src/utils/local-storage.ts',
  /version: '[^']+'/,
  `version: '${newVersion}'`
);

// 7. backend/app/core/config.py
replaceInFile(
  'backend/app/core/config.py',
  /PROJECT_VERSION: str = "[^"]+"/,
  `PROJECT_VERSION: str = "${newVersion}"`
);

// 8. .env.example
replaceInFile(
  '.env.example',
  /PROJECT_VERSION=.*/,
  `PROJECT_VERSION=${newVersion}`
);

// 9. Dockerfile
replaceInFile(
  'Dockerfile',
  /version="[^"]+"/,
  `version="${newVersion}"`
);

// 10. LoginSafe.vue 版权行（硬编码版本号）
replaceInFile(
  'frontend/src/views/auth/LoginSafe.vue',
  /V\d+\.\d+\.\d+/,
  `V${newVersion}`
);

// 11. SystemConfig.vue DEFAULTS
replaceInFile(
  'frontend/src/views/system/SystemConfig.vue',
  /version: 'V[^']+'/,
  `version: 'V${newVersion}'`
);

// 12. generate_ppt.py
replaceInFile(
  'scripts/build/generate_ppt.py',
  /版本 \d+\.\d+\.\d+/g,
  `版本 ${newVersion}`
);
replaceInFile(
  'scripts/build/generate_ppt.py',
  /v\d+\.\d+\.\d+/g,
  `v${newVersion}`
);

console.log(`\n完成: 更新了 ${updatedCount} 个文件\n`);

File: scripts/sync_version.py
"""版本号同步脚本 — 从 version.txt 或 git tag 同步到所有配置文件

用法: python scripts/sync_version.py [version]
  - 不带参数：从 version.txt 读取
  - 带参数：直接使用该版本号 (如 v1.4.0 → 1.4.0)

同步目标: config.py, package.json, NSIS 脚本, .env.example
"""

import json
import os
import re
import sys
from pathlib import Path


PROJECT_ROOT = Path(__file__).resolve().parent.parent


def get_version():
    if len(sys.argv) > 1:
        v = sys.argv[1].lstrip("v")
    else:
        vf = PROJECT_ROOT / "version.txt"
        if vf.exists():
            v = vf.read_text().strip()
        else:
            print("ERROR: 无 version.txt 且未提供参数")
            sys.exit(1)
    return v


def update_config_py(version: str):
    path = PROJECT_ROOT / "backend" / "app" / "core" / "config.py"
    content = path.read_text()
    new_content = re.sub(
        r'PROJECT_VERSION:\s*str\s*=\s*"[^"]*"',
        f'PROJECT_VERSION: str = "{version}"',
        content,
    )
    if new_content != content:
        path.write_text(new_content)
        print(f"  UPD: {path.relative_to(PROJECT_ROOT)}")


def update_package_json(version: str):
    path = PROJECT_ROOT / "frontend" / "package.json"
    pkg = json.loads(path.read_text())
    if pkg.get("version") != version:
        pkg["version"] = version
        path.write_text(json.dumps(pkg, indent=2, ensure_ascii=False) + "\n")
        print(f"  UPD: {path.relative_to(PROJECT_ROOT)}")


def update_nsis_scripts(version: str):
    for nsi in (PROJECT_ROOT / "build-scripts").glob("*.nsi"):
        content = nsi.read_text()
        new_content = re.sub(
            r'!define PRODUCT_VERSION\s+"[^"]*"',
            f'!define PRODUCT_VERSION "{version}"',
            content,
        )
        if new_content != content:
            nsi.write_text(new_content)
            print(f"  UPD: {nsi.relative_to(PROJECT_ROOT)}")


def update_env_example(version: str):
    for env_path in [
        PROJECT_ROOT / "frontend" / ".env.example",
        PROJECT_ROOT / "frontend" / ".env.production",
    ]:
        if not env_path.exists():
            continue
        content = env_path.read_text()
        new_content = re.sub(
            r'VITE_APP_VERSION=\S+',
            f'VITE_APP_VERSION={version}',
            content,
        )
        if new_content != content:
            env_path.write_text(new_content)
            print(f"  UPD: {env_path.relative_to(PROJECT_ROOT)}")


def main():
    version = get_version()
    print(f"Syncing version to {version}...")
    update_config_py(version)
    update_package_json(version)
    update_nsis_scripts(version)
    update_env_example(version)
    print("Done.")


if __name__ == "__main__":
    main()
